JPS_BASE_LIB/src/main/java/uk/ac/cam/cares/jps/base/derivation/DerivationOutputs.java

```diff
@@ -0,0 +1,273 @@
+	public static final String OLD_ENTITIES_DOWNSTREAM_DERIVATION_MAP_ERROR = "Serialise the given JSONObject to oldEntitiesDownstreamDerivationMap Map<String, List<String>> is not supported: ";
+	public static final String OLD_NEW_ENTITIES_MATCHING_ERROR = "When the agent writes new instances, make sure that there is 1 instance with matching rdf:type over the old set, old set: ";
+	public static final String INVALID_IRI_ERROR = "Invalid IRI received when validating IRIs: ";
+	public static final String INVALID_IRI_FOR_GET_CLASS_NAME_ERROR = "The provided rdf:type to getClassName should NOT end with '#' or '/', received IRI: ";
+
+	//////////////////
+	// Constructors //
+	//////////////////
+
+	public DerivationOutputs() {
+		this.oldEntitiesMap = new HashMap<>();
+		this.oldEntitiesDownstreamDerivationMap = new HashMap<>();
+		this.newEntitiesMap = new HashMap<>();
+		this.outputTriples = new ArrayList<TriplePattern>();
+		this.thisDerivation = new String();
+		this.retrievedInputsAt = 0;
+	}
+
+	////////////////////////
+	// Set up information //
+	////////////////////////
+
+	public void setThisDerivation(String derivation) {
+		this.thisDerivation = derivation;
+	}
+
+	public void setRetrievedInputsAt(long retrievedInputsAt) {
+		this.retrievedInputsAt = retrievedInputsAt;
+	}
+
+	public void setOldEntitiesMap(JSONObject oldEntitiesRdfType) {
+		// construct oldEntitiesMap
+		Map<String, String> map = new HashMap<>();
+		Iterator<String> keys = oldEntitiesRdfType.keys();
+		while (keys.hasNext()) {
+			String key = keys.next();
+			Object val = oldEntitiesRdfType.get(key);
+			if (val instanceof String) {
+				map.put(key, (String) val);
+			} else {
+				throw new JPSRuntimeException(OLD_ENTITIES_MAP_ERROR + oldEntitiesRdfType.toString());
+			}
+		}
+		this.oldEntitiesMap = map;
+	}
+
+	public void setOldEntitiesDownstreamDerivationMap(JSONObject oldEntitiesDownstreamDerivation) {
+		// construct oldEntitiesDownstreamDerivationMap
+		Map<String, List<String>> map = new HashMap<>();
+		Iterator<String> keys = oldEntitiesDownstreamDerivation.keys();
+		while (keys.hasNext()) {
+			String key = keys.next();
+			Object val = oldEntitiesDownstreamDerivation.get(key);
+			if (val instanceof JSONArray) {
+				map.put(key, ((JSONArray) val).toList().stream().map(iri -> (String) iri).collect(Collectors.toList()));
+			} else if (val instanceof String) {
+				map.put(key, new ArrayList<>(Arrays.asList((String) val)));
+			} else {
+				throw new JPSRuntimeException(
+						OLD_ENTITIES_DOWNSTREAM_DERIVATION_MAP_ERROR + oldEntitiesDownstreamDerivation.toString());
+			}
+		}
+		this.oldEntitiesDownstreamDerivationMap = map;
+	}
+
+	////////////////////////////////////////////////////////////////////
+	// Create new entities and add triples - to be used by developers //
+	////////////////////////////////////////////////////////////////////
+
+	public void createNewEntity(String iri, String rdfType) {
```

Comment:
The derivation framework does not support multiple IRIs with the same rdf:type, I think an exception should be thrown here if there's a duplicate

Reply:
As we discussed, the framework works fine as long as the outputs are not inputs to another derivation. So I will leave it as it is now.